### PR TITLE
hotel.rb

### PR DESCRIPTION
**Pull request by kayxn23**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | In the beginning I had to decide if certain behaviors should belong to either Room, Reservation or TrackingSystem classes. I considered creating more behaviors in the Reservation class, but eventually decided to write them in the TS class to so that I could centralize all the behaviors, and because it makes more sense for a TrackingSystem to have behaviors than Room and Reservation objects in real life. 
Describe a concept that you gained more clarity on as you worked on this assignment. | I gained more clarity on composition by trial and error. I realize more how mapping out has-a and is-a relationships prior to coding is crucial part of the design process, because each object needs to work together and ideally be flexible for change. Also, I learned that it's difficult to decipher User Stories, and I need to ask more clarification questions around this. 
Describe a _nominal test_ that you wrote for this assignment. | Making sure that the methods throw errors if an invalid argument is entered is in most of my tests. 
Describe an _edge case test_ that you wrote for this assignment. | Making sure that an argument error is raised if a start_date is after an end_date input. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I was able to follow this procedure more than half of the time writing this program. Writing pseudocode comes more naturally than writing tests, but I think this project has made me more comfortable than before with driving my code with Tests. It's difficult to write tests for a method that is more complex and when I'm not sure how it's going to look yet. Writing tests did help guide my code, but I kept getting drawn into writing the code methods when errors came up. I need more practice coming up with edge cases.

### CODE REVIEW

File: lib/reservation.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+class Reservation
+  attr_reader :room_num, :start_time, :end_time, :price
+
+  def initialize(attributes)
```
Comment (CheezItMan) [inline lib/reservation.rb:4]: you should have some checks to make sure the start and end dates are correct here.

File: lib/reservation.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    @start_time = attributes[:start_time]
+    @end_time = attributes[:end_time]
+    @price = attributes[:price]
+  end
+
+  def total_cost
+    total_cost = 0
+    return total_cost = ((end_time - start_time) * price).round(2)
+  end
+
```
Comment (CheezItMan) [inline lib/reservation.rb:15]: I would suggest to add methods to tell if two reservations clash, or if a reservation is on a specific date.

File: lib/tracking_system.rb [hunk 3]
```diff
@@ -0,0 +1,176 @@
+require_relative 'room'
+require_relative 'reservation'
+require_relative 'block'
+require 'pry'
+
+NUMBER_OF_ROOMS = 20
```
Comment (CheezItMan) [inline lib/tracking_system.rb:6]: good use of constants!

File: lib/tracking_system.rb [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+  def view_all_rooms
+    return @all_rooms
+  end
+
+  def view_reservations_on(date)
+    raise ArgumentError.new"#{date} must be instance of Date" unless date.instance_of? Date
+    all_reservations = []
+    @reservations.each do |reservation|
+      if (reservation.start_time...reservation.end_time).include?(date)
```
Comment (CheezItMan) [inline lib/tracking_system.rb]: It would make a great deal of sense to have a method in `Reservation` to tell if the reservation occurs on the given date.  That way the tracking system doesn't need to involve itself in the internal details of a reservation.

File: lib/tracking_system.rb [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  def view_available_rooms_on(start_time: Date.now, end_time: Date.now + 1)
+    raise ArgumentError.new"start_time must be before end_time" unless start_time < end_time
+    available_rooms = []
+    unavailable_count = 0
+    @all_rooms.each do |room|
+      if room.reserved_dates.empty? && room.block == :NA
+        available_rooms << room
+      else
+        room.reserved_dates.each do |dates_hash|
+          if ranges_overlap?((dates_hash[:start_time]...dates_hash[:end_time]).to_a, (start_time..end_time).to_a) == false && room.block == :NA
```
Comment (CheezItMan) [inline lib/tracking_system.rb:50]: Just because a room's reservation range doesn't overlap once, doesn't mean the room isn't reserved on that date.

File: lib/tracking_system.rb [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    raise ArgumentError.new"Not enough rooms available on those dates" if available_rooms.length < number_of_rooms
+    number_of_rooms.times do |i|
+      @reservations << Reservation.new({room_num: available_rooms[i].room_num, start_time: start_time, end_time: end_time, price: 200.00})
+      available_rooms[i].reserved_dates << {start_time: start_time, end_time: end_time}
+    end
+    @reservations
+  end
+
+
+  def add_block(start_time: Date.today + 7, end_time: Date.today.next_month, number_of_rooms: 5, discount: 10)
```
Comment (CheezItMan) [inline lib/tracking_system.rb:77]: This method seems to put a room in a block **once**, if a room is in **any** block it's then unavailable no matter the dates involved.  That's because you're using a `block` field in the `Room` class.

File: lib/tracking_system.rb [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    end #
+  end
+  return discount
+ end
+
+
+private
+
+ def generate_block_id
+  (0..3).map { (65 + rand(26)).chr }.join
```
Comment (CheezItMan) [inline lib/tracking_system.rb:154]: Just want to point out that this method isn't guaranteed to generate unique block ids.

File: lib/tracking_system.rb [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+ end
+
+ def ranges_overlap?(r1, r2)
```
Comment (CheezItMan) [inline lib/tracking_system.rb:157]: 👍

File: spec/tracking_system_spec.rb [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+    end
+
+    it "returns an array" do
+      @reservation = @tracker.add_reservation(start_time: Date.new(2018,10,5), end_time: Date.new(2018,10,10), number_of_rooms:1)
+      @reservation1 = @tracker.add_reservation(start_time: Date.new(2018,10,5), end_time: Date.new(2018,10,10), number_of_rooms:1)
+      expect(@tracker.view_reservations_on(Date.new(2018,10,5))).must_be_kind_of Array
+    end
+
+    it "raises an ArgumentError if no reservations are available on given date" do
+      expect{@tracker.view_reservations_on(Date.new(3333,10,5))}.must_raise ArgumentError
```
Comment (CheezItMan) [inline spec/tracking_system_spec.rb:64]: Why raise an error, why not just return an empty list?

File: spec/tracking_system_spec.rb [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+      expect(@available_rooms).must_be_kind_of Array
+    end
+
+    it "returns an array containing instances of Room" do
+      @tracker = TrackingSystem.new
+      @available_rooms = @tracker.view_available_rooms_on(start_time: Date.new(2018,8,1),end_time:Date.new(2018,9,5))
+      expect(@available_rooms[0]).must_be_kind_of Room
+    end
+
+    it "raises ArgumentError if no rooms are available on these dates(aka they're in a block or reserved)" do
```
Comment (CheezItMan) [inline spec/tracking_system_spec.rb:119]: This test assumes too much detail about how `Room` is implemented.  Instead what you could do is add 20 reservations on a given date using `TrackingSystem`' s method.  Then try making a 21st reservation on the same date range.

File: spec/tracking_system_spec.rb [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+  describe "#add_reservation" do
+    before do
+      @tracker = TrackingSystem.new
+    end
+
+    it "returns an array of reservations" do
+      @reservations = @tracker.add_reservation(start_time: Date.new(2018,8,1), end_time: Date.new(2018,8,25), number_of_rooms:1)
+      @reservations = @tracker.add_reservation(start_time: Date.new(2018,8,1), end_time: Date.new(2018,8,25), number_of_rooms:6)
+      list = @tracker.reservations
+      expect(@reservations).must_be_kind_of Array
```
Comment (CheezItMan) [inline spec/tracking_system_spec.rb:138]: This is more testing `reservations` instead of `add_reservation`

File: spec/tracking_system_spec.rb [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    end
+
+    it "raises ArgumentError if there are not enough available-rooms on the requested date" do
+      expect{@tracker.add_reservation(start_time: Date.new(2018,1,1),end_time:Date.new(2018,1,2),number_of_rooms:350)}.must_raise ArgumentError
+    end
+
```
Comment (CheezItMan) [inline spec/tracking_system_spec.rb:144]: You should also test that if you make 2 reservations in the same date range, they are for different rooms.

Also test that you can make a reservation on the same date as another reservation ends.

File: spec/tracking_system_spec.rb [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+    it "modifys each room's block id to the same block id as its Block" do
+      @block = @tracker.add_block(start_time: Date.new(2018,8,1), end_time: Date.new(2018,8,25), number_of_rooms:5)
+      @blockid = @tracker.blocks[0].block
+      @room_block = @tracker.blocks[0].rooms[0].block
+      expect(@blockid == @room_block).must_equal true
+    end
+  end
+
+  describe "#retrieve_block_dates" do
```
Comment (CheezItMan) [inline spec/tracking_system_spec.rb:189]: I would suggest using a block Id number instead of a symbol just because it's easier to work with.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| Good number of commits and good commit messages
Answer comprehension questions	| Check, I hope you see now that you probably should have broken out some functionality into `Reservation` and `Block`
**Design** |
Each class is responsible for a single piece of the program | You have `TrackingSystem` doing too much.  Pretty much all your code is going there and `Reservation` and `Block` are pretty much just structs.  
Classes are loosely coupled | Yes
**Wave 1** |
List rooms | Check
Reserve a room for a given date range | Check
List reservations for a given date | Check
Calculate reservation price | Check
Invalid date range produces an error | Check, but that check should go in `Block` and `Reservation`
Test coverage | 99%
**Wave 2** |
View available rooms for a given date range | Check
Reserving a room that is not available produces an error | Check
**Wave 3** |
Create a block of rooms | Check
Check if a block has rooms | Check
Reserve a room from a block | Check
**Fundamentals** | 
Names variables, classes and modules appropriately | Check
Understanding of variable scope - local vs instance | Check
Can create complex logical structures utilizing variables | Check
Appropriately uses methods to break down tasks into smaller simpler tasks | Your methods could use some breaking down and your classes are too centralized.  
Understands the differences between class and instance methods | Check
Appropriately uses iterators and enumerables | You mostly just used `.each`.  
Appropriately writes and utilizes classes | Like I've said you centralized too much in `TrackingSystem`.  
Appropriately utilizes modules as a mechanism to encapsulate functionality | No modules
**Wrap Up** |
There is a refactors.txt file | Check
The file provides a roadmap to future changes | Check
**Additional Feedback** | Nice work.  You hit all the learning goals.  I do think your design is a bit awkward as too much is done in `TrackingSystem`, but it all works.  I also left some small comments in your testing.  Overall outstanding work.
